### DISCUSSION TITLE
Can Jotai be used in order to replace the React Context API?

### DISCUSSION THREAD

**Discussion by quangbuule** [pmndrs/jotai#973] [Q&A]
Hi team, I am trying to achieve the React Context behavior, and not sure if we can do that with Jotai. 

Assuming we got a root component that receives data from props and passes down the data to its children. With React Context Provider, it can pass the data to the children in just 1 rendering cycle. I am not sure if we can do that (or if we should do that) with atoms.

Currently, Jotai provider only handles initial values, and there is no API (that I can see) to change the value in 1 render cycle. The only way to work it out is, changing the atom value on side-effect when the root component receives new data.

Am I missing anything? Thanks!

**Comment by dai-shi** [comment]
Hi, our hope is to support all context use cases.
For simple cases, `initialValues` prop should work.
Can you explain what context use case is difficult with it? Maybe a short code snippet?

We do have `useHydrateAtoms` for 1 render cycle, but its usage is limited.

**Reply by quangbuule** [reply]
So, lemme give a quick example here, imaging we are creating a complex button with a text effect, which is can be used like this:
``` tsx
<ComplexButton withTextEffect={true} />
```

The ComplexButton consists of lots of nested and deep inside its children, there is a Text component that needs the `withTextEffect` prop. To pass the `withTextEffect` from the ComplexButton to the Text component, we can use React Context API.


```tsx
const ComplexButton = (props) => (
  <context.Provider value={{ withTextEffect: props.withTextEffect }}>
    {/* asumming there is lots of nested in here */}
    <Text />
    {/* ... */}
  </context.Provider>
);

const Text = () => {
  const { withTextEffect } = useContext(context);
  return ....; // render the text with/without text effect
}
```

Whenever the `ComplexButton` receives the change on the props, it will pass down that `withTextEffect` to the `Text` component in 1 render cycle.

In the case when we want to implement Jotai as replacement of the React Context here. The ComplexButton is supposed to be like this:
```
const ContextButton = (props) => (
  <JotaiProvider initialValues={[ [withTextEffectAtom, props.withTextEffect] ])>
    {/* ... */}
    <Text />
    {/* ... */}
  </JotaiProvider>
)
```
In this case, `withTextEffect` from the `ComplexButton` props takes effect only once. And when it's changed (from the props), the `Text` component won't receive it.

So, I am not sure how we could use Jotai for this situation.

**Reply by dai-shi** [reply]
Thanks for the explanation. Good point.
Jotai atom is modeled after useContext+useState, and in your case, you want to use useContext without useState.
So, this is a little bit tricky, or almost impossible.

You can imagine something like this:
```jsx
const ComplexButton = (props) => {
  const withTextEffectState = useState(props.withTextEffect)
  return (
    <context.Provider value={{ withTextEffectState }}>
      {/* asumming there is lots of nested in here */}
      <Text />
      {/* ... */}
    </context.Provider>
  )
}
```

So, given that jotai atoms are states, there's a mismatch. And, maybe using the bare context is an efficient solution.

If we'd really need to use jotai in this case, we could use `key={}` to remount the entire subtree:
```jsx
const ContextButton = (props) => (
  <JotaiProvider key={String(props.withTextEffect)} initialValues={[ [withTextEffectAtom, props.withTextEffect] ])>
    {/* ... */}
    <Text />
    {/* ... */}
  </JotaiProvider>
)
```

But, I suppose this is not what you want for your case.

---

If you have control on setting `withTextEffect`, you can set an atom value there, instead of passing it as props.

**Reply by quangbuule** [reply]
Thanks a lot for a detailed response! When writing some sort of 3rd components/libraries like the `ComplexButton`, I think, it's pretty common to see the case where we want to provide the context down the tree. Having both Context and Jotai for data passing is not quite ideal.

To accomplish the mission to support all Context use cases, I would like to propose some sort of API to update the atom value on the fly from the Provider, what do you think?

On top of my mind it could be something like:
```tsx
// Should result in readonly atom, cannot set with useAtom
const atom = atom(PRIMITIVE /** initial value */, (updatedValue) => updatedValue); 

<Provider value={[ [atom, value] ]} />
```

**Reply by dai-shi** [reply]
While we never want to extend the atom api in such a way, this is not the issue of the api design.
How would you be able to solve your case with useContext+useState?

We don't pass atom values through context. It's store and subscription based.
So, the right answer is we should accept two renders if we need to sync prop value and atom value.

> To accomplish the mission to support all Context use cases

Meaning, we should say our mission is to support the use cases of context + state. 😓

**Reply by quangbuule** [reply]
Thanks @dai-shi, seems like, with context + state, it looks like a stateful component that is tricky to get derived state from props. Maybe the only way is to trigger another render.

**Reply by TwistedMinda** [reply]
Can we guarentee that the view hierarchy is read line-by-line ?
If yes we could be able to do something like

```typescript
const activeAtom = atomFamily(({key, defaultValue}) => atom(defaultValue), deepEqual)

var lastKey = ''
const ComplexButton = (props) => {
  lastKey = props.key
  const setActive = useSetAtom(activeAtom({key: lastKey, defaultValue: props.active}))
  React.useEffect(() => { setActive(props.active) }, [props.active])
  console.log(lastKey, active)
  return (
    <BigNesting>
      <Text>
    </BigNesting>
  )
}

const Text = () => {
  const active = useAtomValue(activeAtom({key: lastKey, defaultValue: false}))
  console.log(lastKey, active)
  return (null)
}
```
Lol.
I'm genuinely intrigued that there is no way to do this cleanly with jotai without react context!

**Reply by dai-shi** [reply]
> I'm genuinely intrigued that there is no way to do this cleanly with jotai without react context!

jotai atoms in react context (or props) is possible and a valid pattern.
we suggest atoms-in-atom pattern in such a case.

**Reply by TwistedMinda** [reply]
@dai-shi Could you provide some pseudo code to show the idea? I'm really not understanding how u guys fix this issue.

If there is only one ComplexButton in the app its easy, cause you can use a single atom.
But if there are many ComplexButton's, and you decide to pass an atom via prop, how do you access this atom from the nested <Text>  component ?

Thank you for your time

**Reply by dai-shi** [reply]
Does this doc help? https://jotai.org/docs/guides/atoms-in-atom
tbh, I'm not confident if I understand your issue correctly.

**Reply by TwistedMinda** [reply]
@dai-shi Hm. No I don't understand how you fix OP's issue with atoms as props. That would mean to me that the atom prop would have to be passed down every child component in our "ComplexButton - Lots of nesting - Text" Example.

like so:
```typescript
const ComplexButton = (props) => {
  return (<SubView atomProp={props.atomProp}></SubView />)
}
const Subview = (props) => {
  return (<MoreNestedView atomProp={props.atomProp}></MoreNestedView />)
}
const MoreNestedView = (props) => {
  return (<Text atomProp={props.atomProp}></Text />)
}
const Text = (props) => {
  const propValue = useAtomValue(props.atomProp}
  console.log('yay we got it', propValue)
  return (null)
}
```
Doesn't look good at all, why would even use an atom here instead of a normal prop.
I'm lost :p

**Reply by dai-shi** [reply]
Sorry, I think I misunderstood your comment in the last line of https://github.com/pmndrs/jotai/discussions/973#discussioncomment-2764737. Please forget about atoms-in-atom thing.

As far as I remember, the OP's issue is if we can use jotai for context without state. Then, no. It's impossible, and out of the scope. jotai doesn't do prop propagation, but does state subscription.
